src/design_notes/copy_ergonomics.md

```diff
@@ -0,0 +1,105 @@
+primarily a performance concern, so the problem is usually that these copies are
+easy to miss. It is also true that linting on these cases is really difficult,
+as depending on the problem, it can be quite reasonable to produce
+megabyte-large copies once per program, but in a web server doing so on each
+request may not be a good idea.
+
+Implementations of `Copy` on closures and arrays are the prime example of Rust
+currently being overeager with the defaults in some contexts.
+
+This also comes up with `Copy` impls on `Range`, which would generally be
+desirable but is error-prone given the `Iterator/IntoIterator` impls on ranges.
+
+The example here does not compile today (since Range is not Copy), but would be
+unintuitive if it did.
+
+```rust,compile_fail
+let mut x = 0..10;
+let mut c = move || x.next();
+println!("{:?}", x.next()); // prints 0
+println!("{:?}", c()); // prints 0, because the captured x is implicitly copied.
+```
+
+This example illustrates the range being copied into the closure, while the user
+likely expected the name "x" to refer to the same range in both cases.
+
+A lint has been [proposed](https://github.com/rust-lang/rust/issues/45683) to
+permit Copy impls on types where Copy is likely not desirable with particular
+conditions (e.g., Copy of IntoIterator-implementing types after iteration).
+
+Note that "large copies" comes up with moves as well (which are copies, just
+taking ownership as well), so a size-based lint is plausibly desirable for both.
+
+### History
+
+* Proposed lint: [#45683](https://github.com/rust-lang/rust/issues/45683)
+
+## References to `Copy` types
+
+Frequently when dealing with code generic over T you end up needing things like
+`[u8]::contains(&5)` which is ugly and annoying. Iterators of copy types also
+produce `&&u64` and similar constructs which can produce unexpected type errors.
+
+```rust
+for x in &vec![1, 2, 3, 4, 5, 6, 7] {
+    process(*x); // <-- annoying that we need `*x`
+}
+
+fn process(x: i32) { }
+```
+
+```rust
+fn sum_even(v: &[u32]) -> u32 {
+    // **v is annoying
+    v.iter().filter(|v| **v % 2 == 0).sum()
+}
+```
+
+Note that this means that you in most cases want to "boil down" to the inner
+type when dealing with references, i.e., `&&u32` you actually want `u32`, not
+`&u32`. Notably, though, this may *not* be true if the Copy type is something
+more complex (e.g., a future Copy Cell), since then `&Cell` is quite different
+from a `Cell`, the latter being likely useless for modification at least.
+
+There is also plausibly performance left on the table with types like `&&u64`.
+
+Note that this interacts with the unintentional copies (especially of large
+structures).
+
+This could plausibly be done with moved values as well, so long as the
+semantics match the syntax (e.g. `wants_ref(foo)` acts like `wants_ref(&{foo})`)
+similar to how one can pass `&mut` to something that only wants `&`.
```

Comment:
```suggestion
similar to how one can pass `&mut` to something that only wants `&`.
This would be a tradeoff: in some cases people may want the type-checker to flag such cases and require explicitly taking a reference, while in other cases people may want the compiler to automatically make such code work. We would want to consider and evaluate this tradeoff, and whether we can usefully separate such cases.
```